File: Graph/problem399/problem399.cpp
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    vector<double> calcEquation(
        vector<vector<string>>& equations,
        vector<double>& values,
        vector<vector<string>>& queries
    ) {
        unordered_map<string, vector<pair<string, double>>> graph;

        for (int i = 0; i < static_cast<int>(equations.size()); ++i) {
            const string& a = equations[i][0];
            const string& b = equations[i][1];
            double v = values[i];
            graph[a].push_back({b, v});
            graph[b].push_back({a, 1.0 / v});
        }

        auto bfs = [&](const string& src, const string& dst) -> double {
            if (!graph.count(src) || !graph.count(dst)) return -1.0;
            if (src == dst) return 1.0;

            unordered_set<string> visited;
            queue<pair<string, double>> q;
            q.push({src, 1.0});
            visited.insert(src);

            while (!q.empty()) {
                auto [node, prod] = q.front();
                q.pop();
                for (const auto& [nb, w] : graph[node]) {
                    if (nb == dst) return prod * w;
                    if (!visited.count(nb)) {
                        visited.insert(nb);
                        q.push({nb, prod * w});
                    }
                }
            }
            return -1.0;
        };

        vector<double> ans;
        ans.reserve(queries.size());
        for (const auto& q : queries) {
            ans.push_back(bfs(q[0], q[1]));
        }
        return ans;
    }
};

File: Graph/problem399/problem399.py
from collections import defaultdict, deque
from typing import List


class Solution:
    def calcEquation(
        self,
        equations: List[List[str]],
        values: List[float],
        queries: List[List[str]],
    ) -> List[float]:
        graph: dict[str, list[tuple[str, float]]] = defaultdict(list)

        for (a, b), v in zip(equations, values):
            graph[a].append((b, v))
            graph[b].append((a, 1.0 / v))

        def bfs(src: str, dst: str) -> float:
            if src not in graph or dst not in graph:
                return -1.0
            if src == dst:
                return 1.0
            visited = {src}
            queue = deque([(src, 1.0)])
            while queue:
                node, prod = queue.popleft()
                for nb, w in graph[node]:
                    if nb == dst:
                        return prod * w
                    if nb not in visited:
                        visited.add(nb)
                        queue.append((nb, prod * w))
            return -1.0

        return [bfs(c, d) for c, d in queries]
